File: ios/BetterLise/Core/API/LiseHealthMonitor.swift
import Foundation
import Observation

/// Tracks how fast Lise currently answers so loading screens can pace themselves on reality.
@MainActor
@Observable
final class LiseHealthMonitor {
    nonisolated static let defaultDuration: TimeInterval = 10
    private nonisolated static let minimumSamples = 4
    private static let refreshInterval: TimeInterval = 5 * 60
    private nonisolated static let slowThreshold: TimeInterval = 15

    private(set) var health: LiseHealth?
    private var fetchedAt: Date?
    private let session: SessionStore

    init(session: SessionStore) {
        self.session = session
    }

    var expectedSyncDuration: TimeInterval { Self.expectedDuration(for: health) }
    var slowNotice: String? { Self.slowNotice(for: health) }

    func refreshIfNeeded(now: Date = Date()) async {
        if let fetchedAt, now.timeIntervalSince(fetchedAt) < Self.refreshInterval { return }
        fetchedAt = now
        // Pacing is cosmetic: keep the previous estimate if the server can't be reached
        if let health = try? await session.sendPublic(Endpoints.health()) {
            self.health = health
        }
    }

    nonisolated static func expectedDuration(for health: LiseHealth?) -> TimeInterval {
        guard let health, health.count >= minimumSamples, health.avgDuration > 0 else { return defaultDuration }
        return min(max(health.avgDuration / 1000, 3), 60)
    }

    nonisolated static func slowNotice(for health: LiseHealth?) -> String? {
        guard let health, health.count >= minimumSamples else { return nil }
        let seconds = health.avgDuration / 1000
        guard seconds > slowThreshold else { return nil }
        return "Lise est lente en ce moment (≈ \(Int(seconds)) s par synchronisation)."
    }
}

File: android/app/src/main/java/com/betterlise/app/data/health/LiseHealthMonitor.kt
package com.betterlise.app.data.health

import com.betterlise.app.data.api.LiseHealth
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

/** Tracks how fast Lise currently answers so loading screens can pace themselves on reality. */
class LiseHealthMonitor(
    private val fetch: suspend () -> LiseHealth,
    private val clock: () -> Long = System::currentTimeMillis,
) {
    private val _health = MutableStateFlow<LiseHealth?>(null)
    val health: StateFlow<LiseHealth?> = _health.asStateFlow()
    private var fetchedAt: Long? = null

    suspend fun refreshIfNeeded() {
        val now = clock()
        fetchedAt?.let { if (now - it < REFRESH_INTERVAL_MS) return }
        fetchedAt = now
        // Pacing is cosmetic: keep the previous estimate if the server can't be reached
        try {
            _health.value = fetch()
        } catch (e: CancellationException) {
            throw e
        } catch (e: Exception) {
            return
        }
    }

    companion object {
        const val DEFAULT_DURATION_SECONDS = 10.0
        private const val MINIMUM_SAMPLES = 4
        private const val REFRESH_INTERVAL_MS = 5 * 60 * 1000L
        private const val SLOW_THRESHOLD_SECONDS = 15.0

        fun expectedDurationSeconds(health: LiseHealth?): Double {
            if (health == null || health.count < MINIMUM_SAMPLES || health.avgDuration <= 0) return DEFAULT_DURATION_SECONDS
            return (health.avgDuration / 1000).coerceIn(3.0, 60.0)
        }

        fun slowNotice(health: LiseHealth?): String? {
            if (health == null || health.count < MINIMUM_SAMPLES) return null
            val seconds = health.avgDuration / 1000
            if (seconds <= SLOW_THRESHOLD_SECONDS) return null
            return "Lise est lente en ce moment (≈ ${seconds.toInt()} s par synchronisation)."
        }
    }
}
